bokehjs/src/lib/models/annotations/tooltip.ts

```diff
@@ -89,7 +91,7 @@ export class TooltipView extends AnnotationView {
         break
       case "left":
         this.el.classList.add(bk_right)
-        left = sx - this.el.offsetWidth - arrow_size
+        right = (this.ctx.canvas.width - sx) + arrow_size
```

Comment:
@jsignell the other three cases still use e.g. `el.offsetXYZ` would it be possible/preferable to update them correspondingly to use `ctx.canvas.XYZ` for consistency?

Reply:
Possibly. I was trying to do the minimal fix and I wasn't positive that using ctx.canvas would be desirable. I can make the others follow the new pattern if you like it.

Reply:
This gets called many many times, so it probably matters for performance, but I'm not sure how to test that.

Reply:
@mattpap any comments on this? I'm not sure reading canvas dimensions incurs appreciable cost. I'm not opposed to a minimal fix either but then maybe a new issue to investigate making them consistent (because months/years later that kind of inconsistency is a head-scratcher for maintenance)

Reply:
Honestly, it's not clear to me how this code is supposed to work. AFAICT `offsetWidth` and `clientWidth` are computed and used to set `left` and `top`, but then they can change based on the space available.

Reply:
I don't know, most likely not. However, I don't see why one would use `ctx.canvas`. That is pretty much private information from tooltip's perspective. Using layout would be much better (`this.plot_view.layout.bbox`).

Reply:
Cool, I didn't notice that the `bbox` there was the same as the `canvas.width`.

Reply:
I think I'd prefer to merge this bit of functionality and change the rest down the road. If that is ok with you @bryevdv and @mattpap